File: apps/backend/app/api/routers/expenses.py
from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Optional

from app.core.auth import require_user_id
from app.core.tenancy import assert_org_member, assert_org_role
from app.db.supabase import get_supabase_client
from app.schemas.domain import CreateExpenseInput, UpdateExpenseInput
from app.services.audit import write_audit_log
from app.services.enrichment import enrich_expenses
from app.services.fx import get_usd_to_pyg_rate
from app.services.table_service import create_row, delete_row, get_row, update_row

router = APIRouter(tags=["Expenses"])


@router.get("/expenses")
def list_expenses(
    org_id: str = Query(...),
    from_date: Optional[str] = Query(None, alias="from"),
    to_date: Optional[str] = Query(None, alias="to"),
    category: Optional[str] = Query(None),
    currency: Optional[str] = Query(None),
    payment_method: Optional[str] = Query(None),
    vendor_name: Optional[str] = Query(None),
    property_id: Optional[str] = Query(None),
    unit_id: Optional[str] = Query(None),
    reservation_id: Optional[str] = Query(None),
    limit: int = Query(300, ge=1, le=2000),
    user_id: str = Depends(require_user_id),
) -> dict:
    assert_org_member(user_id, org_id)

    try:
        client = get_supabase_client()
        query = client.table("expenses").select("*").eq("organization_id", org_id)

        if category:
            query = query.eq("category", category)
        if currency:
            query = query.eq("currency", currency)
        if payment_method:
            query = query.eq("payment_method", payment_method)
        if property_id:
            query = query.eq("property_id", property_id)
        if unit_id:
            query = query.eq("unit_id", unit_id)
        if reservation_id:
            query = query.eq("reservation_id", reservation_id)
        if from_date:
            query = query.gte("expense_date", from_date)
        if to_date:
            query = query.lte("expense_date", to_date)
        if vendor_name:
            query = query.ilike("vendor_name", f"%{vendor_name.strip()}%")

        # Pagination: offset/limit isn't exposed on this endpoint yet. Keep it simple.
        query = query.order("expense_date", desc=True).range(0, limit - 1)
        response = query.execute()
        rows = response.data or []
    except HTTPException:
        raise
    except Exception as exc:  # pragma: no cover - framework guard
        raise HTTPException(status_code=502, detail=f"Supabase request failed: {exc}") from exc

    return {"data": enrich_expenses(rows, org_id)}


@router.post("/expenses", status_code=201)
def create_expense(payload: CreateExpenseInput, user_id: str = Depends(require_user_id)) -> dict:
    assert_org_role(user_id, payload.organization_id, {"owner_admin", "accountant"})

    record = payload.model_dump(exclude_none=True)
    record["created_by_user_id"] = user_id

    receipt_url = str(record.get("receipt_url") or "").strip()
    if not receipt_url:
        raise HTTPException(status_code=400, detail="receipt_url is required.")
    record["receipt_url"] = receipt_url

    reservation_id = str(record.get("reservation_id") or "").strip()
    if reservation_id:
        reservation = get_row("reservations", reservation_id)
        if reservation.get("organization_id") != payload.organization_id:
            raise HTTPException(status_code=400, detail="reservation_id does not belong to this organization.")

        unit_id = reservation.get("unit_id")
        if not isinstance(unit_id, str) or not unit_id.strip():
            raise HTTPException(status_code=400, detail="reservation_id is missing unit_id.")
        record["unit_id"] = unit_id

        unit = get_row("units", unit_id)
        if unit.get("organization_id") != payload.organization_id:
            raise HTTPException(status_code=400, detail="reservation unit does not belong to this organization.")
        property_id = unit.get("property_id")
        if isinstance(property_id, str) and property_id.strip():
            record["property_id"] = property_id

    currency = str(record.get("currency") or "PYG").strip().upper()
    record["currency"] = currency
    if currency == "USD":
        fx_rate = record.get("fx_rate_to_pyg")
        if fx_rate is None:
            fetched = get_usd_to_pyg_rate(payload.expense_date)
            if fetched is None:
                raise HTTPException(
                    status_code=400,
                    detail="fx_rate_to_pyg is required for USD expenses (auto-fetch failed).",
                )
            record["fx_rate_to_pyg"] = fetched
    else:
        # Keep data consistent: we only store fx_rate_to_pyg for USD expenses.
        record.pop("fx_rate_to_pyg", None)

    created = create_row("expenses", record)
    write_audit_log(
        organization_id=payload.organization_id,
        actor_user_id=user_id,
        action="create",
        entity_name="expenses",
        entity_id=created.get("id"),
        before_state=None,
        after_state=created,
    )
    return enrich_expenses([created], payload.organization_id)[0]


@router.get("/expenses/{expense_id}")
def get_expense(expense_id: str, user_id: str = Depends(require_user_id)) -> dict:
    record = get_row("expenses", expense_id)
    assert_org_member(user_id, record["organization_id"])
    return enrich_expenses([record], record["organization_id"])[0]


@router.patch("/expenses/{expense_id}")
def update_expense(expense_id: str, payload: UpdateExpenseInput, user_id: str = Depends(require_user_id)) -> dict:
    record = get_row("expenses", expense_id)
    assert_org_role(user_id, record["organization_id"], {"owner_admin", "accountant"})

    patch = payload.model_dump(exclude_none=True)

    if "receipt_url" in patch:
        receipt_url = str(patch.get("receipt_url") or "").strip()
        if not receipt_url:
            raise HTTPException(status_code=400, detail="receipt_url cannot be empty.")
        patch["receipt_url"] = receipt_url

    reservation_id = str(patch.get("reservation_id") or "").strip() if "reservation_id" in patch else ""
    if reservation_id:
        reservation = get_row("reservations", reservation_id)
        if reservation.get("organization_id") != record["organization_id"]:
            raise HTTPException(status_code=400, detail="reservation_id does not belong to this organization.")

        unit_id = reservation.get("unit_id")
        if not isinstance(unit_id, str) or not unit_id.strip():
            raise HTTPException(status_code=400, detail="reservation_id is missing unit_id.")
        patch["unit_id"] = unit_id

        unit = get_row("units", unit_id)
        if unit.get("organization_id") != record["organization_id"]:
            raise HTTPException(status_code=400, detail="reservation unit does not belong to this organization.")
        property_id = unit.get("property_id")
        if isinstance(property_id, str) and property_id.strip():
            patch["property_id"] = property_id

    currency = None
    if "currency" in patch:
        currency = str(patch.get("currency") or "").strip().upper()
        if not currency:
            raise HTTPException(status_code=400, detail="currency cannot be empty.")
        patch["currency"] = currency

    effective_currency = currency or str(record.get("currency") or "PYG").strip().upper()
    effective_date = str(patch.get("expense_date") or record.get("expense_date") or "").strip()

    if effective_currency == "USD":
        if "fx_rate_to_pyg" not in patch and record.get("fx_rate_to_pyg") is None:
            fetched = get_usd_to_pyg_rate(effective_date)
            if fetched is None:
                raise HTTPException(
                    status_code=400,
                    detail="fx_rate_to_pyg is required for USD expenses (auto-fetch failed).",
                )
            patch["fx_rate_to_pyg"] = fetched
    else:
        # Ensure non-USD expenses don't keep stale FX snapshots.
        patch["fx_rate_to_pyg"] = None

    updated = update_row("expenses", expense_id, patch)
    write_audit_log(
        organization_id=record.get("organization_id"),
        actor_user_id=user_id,
        action="update",
        entity_name="expenses",
        entity_id=expense_id,
        before_state=record,
        after_state=updated,
    )
    return enrich_expenses([updated], record["organization_id"])[0]


@router.delete("/expenses/{expense_id}")
def delete_expense(expense_id: str, user_id: str = Depends(require_user_id)) -> dict:
    record = get_row("expenses", expense_id)
    assert_org_role(user_id, record["organization_id"], {"owner_admin", "accountant"})
    deleted = delete_row("expenses", expense_id)
    write_audit_log(
        organization_id=record.get("organization_id"),
        actor_user_id=user_id,
        action="delete",
        entity_name="expenses",
        entity_id=expense_id,
        before_state=deleted,
        after_state=None,
    )
    return deleted

File: apps/backend-rs/src/routes/expenses.rs
use axum::{
    extract::{Path, Query, State},
    http::HeaderMap,
    response::IntoResponse,
    Json,
};
use serde_json::{json, Map, Value};
use sqlx::{Postgres, QueryBuilder, Row};

use crate::{
    auth::require_user_id,
    error::{AppError, AppResult},
    repository::table_service::{create_row, delete_row, get_row, update_row},
    schemas::{
        clamp_limit_in_range, remove_nulls, serialize_to_map, CreateExpenseInput,
        ExpenseApprovalInput, ExpensePath, ExpensesQuery, UpdateExpenseInput,
    },
    services::{audit::write_audit_log, enrichment::enrich_expenses, fx::get_usd_to_pyg_rate},
    state::AppState,
    tenancy::{assert_org_member, assert_org_role},
};

pub fn router() -> axum::Router<AppState> {
    axum::Router::new()
        .route(
            "/expenses",
            axum::routing::get(list_expenses).post(create_expense),
        )
        .route(
            "/expenses/{expense_id}/approve",
            axum::routing::post(approve_expense),
        )
        .route(
            "/expenses/{expense_id}/reject",
            axum::routing::post(reject_expense),
        )
        .route(
            "/expenses/{expense_id}",
            axum::routing::get(get_expense)
                .patch(update_expense)
                .delete(delete_expense),
        )
}

async fn list_expenses(
    State(state): State<AppState>,
    Query(query): Query<ExpensesQuery>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    let user_id = require_user_id(&state, &headers).await?;
    assert_org_member(&state, &user_id, &query.org_id).await?;
    let pool = db_pool(&state)?;

    let rows = list_expense_rows(pool, &query).await?;
    let enriched = enrich_expenses(pool, rows, &query.org_id).await?;
    Ok(Json(json!({ "data": enriched })))
}

async fn create_expense(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(payload): Json<CreateExpenseInput>,
) -> AppResult<impl IntoResponse> {
    let user_id = require_user_id(&state, &headers).await?;
    assert_org_role(
        &state,
        &user_id,
        &payload.organization_id,
        &["owner_admin", "accountant"],
    )
    .await?;
    let pool = db_pool(&state)?;

    let mut record = remove_nulls(serialize_to_map(&payload));
    record.insert(
        "created_by_user_id".to_string(),
        Value::String(user_id.clone()),
    );

    let receipt_url = string_from_map(&record, "receipt_url").unwrap_or_default();
    if receipt_url.is_empty() {
        return Err(AppError::BadRequest("receipt_url is required.".to_string()));
    }
    record.insert("receipt_url".to_string(), Value::String(receipt_url));

    let reservation_id = string_from_map(&record, "reservation_id").unwrap_or_default();
    if !reservation_id.is_empty() {
        attach_reservation_unit_property(
            pool,
            &payload.organization_id,
            &reservation_id,
            &mut record,
        )
        .await?;
    }

    let currency = string_from_map(&record, "currency")
        .filter(|value| !value.is_empty())
        .unwrap_or_else(|| "PYG".to_string())
        .to_uppercase();
    record.insert("currency".to_string(), Value::String(currency.clone()));

    if currency == "USD" {
        if !record.contains_key("fx_rate_to_pyg") {
            let fetched =
                get_usd_to_pyg_rate(&state.http_client, payload.expense_date.trim()).await;
            if let Some(rate) = fetched {
                record.insert("fx_rate_to_pyg".to_string(), json!(rate));
            } else {
                return Err(AppError::BadRequest(
                    "fx_rate_to_pyg is required for USD expenses (auto-fetch failed).".to_string(),
                ));
            }
        }
    } else {
        record.remove("fx_rate_to_pyg");
    }

    // IVA auto-calculation (10% in Paraguay)
    if payload.iva_applicable {
        let iva = payload.iva_amount.unwrap_or_else(|| {
            let base = payload.amount;
            (base * 0.10 * 100.0).round() / 100.0
        });
        record.insert("iva_amount".to_string(), json!(iva));
        record.insert("iva_applicable".to_string(), json!(true));
    }

    // Auto-categorize if no category provided
    let has_category = record
        .get("category")
        .and_then(Value::as_str)
        .is_some_and(|s| !s.trim().is_empty());
    if !has_category {
        let vendor = string_from_map(&record, "vendor").unwrap_or_default();
        let desc = string_from_map(&record, "description").unwrap_or_default();
        if let Some(category) =
            crate::services::expense_categorization::auto_categorize(&vendor, &desc, payload.amount)
        {
            record.insert("category".to_string(), Value::String(category.to_string()));
        }
    }

    let created = create_row(pool, "expenses", &record).await?;
    let entity_id = value_str(&created, "id");

    write_audit_log(
        state.db_pool.as_ref(),
        Some(&payload.organization_id),
        Some(&user_id),
        "create",
        "expenses",
        Some(&entity_id),
        None,
        Some(created.clone()),
    )
    .await;

    let mut enriched = enrich_expenses(pool, vec![created], &payload.organization_id).await?;
    Ok((
        axum::http::StatusCode::CREATED,
        Json(enriched.pop().unwrap_or_else(|| Value::Object(Map::new()))),
    ))
}

async fn get_expense(
    State(state): State<AppState>,
    Path(path): Path<ExpensePath>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    let user_id = require_user_id(&state, &headers).await?;
    let pool = db_pool(&state)?;

    let record = get_row(pool, "expenses", &path.expense_id, "id").await?;
    let org_id = value_str(&record, "organization_id");
    assert_org_member(&state, &user_id, &org_id).await?;

    let mut enriched = enrich_expenses(pool, vec![record], &org_id).await?;
    Ok(Json(
        enriched.pop().unwrap_or_else(|| Value::Object(Map::new())),
    ))
}

async fn update_expense(
    State(state): State<AppState>,
    Path(path): Path<ExpensePath>,
    headers: HeaderMap,
    Json(payload): Json<UpdateExpenseInput>,
) -> AppResult<Json<Value>> {
    let user_id = require_user_id(&state, &headers).await?;
    let pool = db_pool(&state)?;

    let record = get_row(pool, "expenses", &path.expense_id, "id").await?;
    let org_id = value_str(&record, "organization_id");
    assert_org_role(&state, &user_id, &org_id, &["owner_admin", "accountant"]).await?;

    let mut patch = remove_nulls(serialize_to_map(&payload));

    if patch.contains_key("receipt_url") {
        let receipt_url = string_from_map(&patch, "receipt_url").unwrap_or_default();
        if receipt_url.is_empty() {
            return Err(AppError::BadRequest(
                "receipt_url cannot be empty.".to_string(),
            ));
        }
        patch.insert("receipt_url".to_string(), Value::String(receipt_url));
    }

    if patch.contains_key("reservation_id") {
        let reservation_id = string_from_map(&patch, "reservation_id").unwrap_or_default();
        if !reservation_id.is_empty() {
            attach_reservation_unit_property(pool, &org_id, &reservation_id, &mut patch).await?;
        }
    }

    let mut currency: Option<String> = None;
    if patch.contains_key("currency") {
        let next_currency = string_from_map(&patch, "currency")
            .unwrap_or_default()
            .to_uppercase();
        if next_currency.is_empty() {
            return Err(AppError::BadRequest(
                "currency cannot be empty.".to_string(),
            ));
        }
        patch.insert("currency".to_string(), Value::String(next_currency.clone()));
        currency = Some(next_currency);
    }

    let effective_currency = currency
        .clone()
        .or_else(|| string_from_value(record.get("currency")))
        .unwrap_or_else(|| "PYG".to_string())
        .to_uppercase();
    let effective_date = string_from_map(&patch, "expense_date")
        .or_else(|| string_from_value(record.get("expense_date")))
        .unwrap_or_default();

    if effective_currency == "USD" {
        if !patch.contains_key("fx_rate_to_pyg") && record.get("fx_rate_to_pyg").is_none() {
            let fetched = get_usd_to_pyg_rate(&state.http_client, effective_date.trim()).await;
            if let Some(rate) = fetched {
                patch.insert("fx_rate_to_pyg".to_string(), json!(rate));
            } else {
                return Err(AppError::BadRequest(
                    "fx_rate_to_pyg is required for USD expenses (auto-fetch failed).".to_string(),
                ));
            }
        }
    } else {
        patch.insert("fx_rate_to_pyg".to_string(), Value::Null);
    }

    let updated = update_row(pool, "expenses", &path.expense_id, &patch, "id").await?;

    write_audit_log(
        state.db_pool.as_ref(),
        Some(&org_id),
        Some(&user_id),
        "update",
        "expenses",
        Some(&path.expense_id),
        Some(record),
        Some(updated.clone()),
    )
    .await;

    let mut enriched = enrich_expenses(pool, vec![updated], &org_id).await?;
    Ok(Json(
        enriched.pop().unwrap_or_else(|| Value::Object(Map::new())),
    ))
}

async fn delete_expense(
    State(state): State<AppState>,
    Path(path): Path<ExpensePath>,
    headers: HeaderMap,
) -> AppResult<Json<Value>> {
    let user_id = require_user_id(&state, &headers).await?;
    let pool = db_pool(&state)?;

    let record = get_row(pool, "expenses", &path.expense_id, "id").await?;
    let org_id = value_str(&record, "organization_id");
    assert_org_role(&state, &user_id, &org_id, &["owner_admin", "accountant"]).await?;

    let deleted = delete_row(pool, "expenses", &path.expense_id, "id").await?;

    write_audit_log(
        state.db_pool.as_ref(),
        Some(&org_id),
        Some(&user_id),
        "delete",
        "expenses",
        Some(&path.expense_id),
        Some(deleted.clone()),
        None,
    )
    .await;

    Ok(Json(deleted))
}

async fn approve_expense(
    State(state): State<AppState>,
    Path(path): Path<ExpensePath>,
    headers: HeaderMap,
    Json(_payload): Json<ExpenseApprovalInput>,
) -> AppResult<Json<Value>> {
    let user_id = require_user_id(&state, &headers).await?;
    let pool = db_pool(&state)?;

    let record = get_row(pool, "expenses", &path.expense_id, "id").await?;
    let org_id = value_str(&record, "organization_id");
    assert_org_role(&state, &user_id, &org_id, &["owner_admin"]).await?;

    let mut patch = Map::new();
    patch.insert(
        "approval_status".to_string(),
        Value::String("approved".to_string()),
    );
    patch.insert("approved_by".to_string(), Value::String(user_id.clone()));
    patch.insert(
        "approved_at".to_string(),
        Value::String(chrono::Utc::now().to_rfc3339()),
    );

    let updated = update_row(pool, "expenses", &path.expense_id, &patch, "id").await?;

    write_audit_log(
        state.db_pool.as_ref(),
        Some(&org_id),
        Some(&user_id),
        "approve",
        "expenses",
        Some(&path.expense_id),
        Some(record),
        Some(updated.clone()),
    )
    .await;

    let mut enriched = enrich_expenses(pool, vec![updated], &org_id).await?;
    Ok(Json(
        enriched.pop().unwrap_or_else(|| Value::Object(Map::new())),
    ))
}

async fn reject_expense(
    State(state): State<AppState>,
    Path(path): Path<ExpensePath>,
    headers: HeaderMap,
    Json(_payload): Json<ExpenseApprovalInput>,
) -> AppResult<Json<Value>> {
    let user_id = require_user_id(&state, &headers).await?;
    let pool = db_pool(&state)?;

    let record = get_row(pool, "expenses", &path.expense_id, "id").await?;
    let org_id = value_str(&record, "organization_id");
    assert_org_role(&state, &user_id, &org_id, &["owner_admin"]).await?;

    let mut patch = Map::new();
    patch.insert(
        "approval_status".to_string(),
        Value::String("rejected".to_string()),
    );
    patch.insert("approved_by".to_string(), Value::String(user_id.clone()));
    patch.insert(
        "approved_at".to_string(),
        Value::String(chrono::Utc::now().to_rfc3339()),
    );

    let updated = update_row(pool, "expenses", &path.expense_id, &patch, "id").await?;

    write_audit_log(
        state.db_pool.as_ref(),
        Some(&org_id),
        Some(&user_id),
        "reject",
        "expenses",
        Some(&path.expense_id),
        Some(record),
        Some(updated.clone()),
    )
    .await;

    let mut enriched = enrich_expenses(pool, vec![updated], &org_id).await?;
    Ok(Json(
        enriched.pop().unwrap_or_else(|| Value::Object(Map::new())),
    ))
}

async fn list_expense_rows(pool: &sqlx::PgPool, query: &ExpensesQuery) -> AppResult<Vec<Value>> {
    let mut builder = QueryBuilder::<Postgres>::new(
        "SELECT row_to_json(t) AS row FROM expenses t WHERE organization_id = ",
    );
    builder.push_bind(query.org_id.clone());

    if let Some(category) = non_empty_opt(query.category.as_deref()) {
        builder.push(" AND category = ");
        builder.push_bind(category);
    }
    if let Some(currency) = non_empty_opt(query.currency.as_deref()) {
        builder.push(" AND currency = ");
        builder.push_bind(currency);
    }
    if let Some(payment_method) = non_empty_opt(query.payment_method.as_deref()) {
        builder.push(" AND payment_method = ");
        builder.push_bind(payment_method);
    }
    if let Some(property_id) = non_empty_opt(query.property_id.as_deref()) {
        builder.push(" AND property_id = ");
        builder.push_bind(property_id);
    }
    if let Some(unit_id) = non_empty_opt(query.unit_id.as_deref()) {
        builder.push(" AND unit_id = ");
        builder.push_bind(unit_id);
    }
    if let Some(reservation_id) = non_empty_opt(query.reservation_id.as_deref()) {
        builder.push(" AND reservation_id = ");
        builder.push_bind(reservation_id);
    }
    if let Some(from_date) = non_empty_opt(query.from_date.as_deref()) {
        builder.push(" AND expense_date >= ");
        builder.push_bind(from_date);
    }
    if let Some(to_date) = non_empty_opt(query.to_date.as_deref()) {
        builder.push(" AND expense_date <= ");
        builder.push_bind(to_date);
    }
    if let Some(vendor_name) = non_empty_opt(query.vendor_name.as_deref()) {
        builder.push(" AND vendor_name ILIKE ");
        builder.push_bind(format!("%{vendor_name}%"));
    }
    if let Some(approval_status) = non_empty_opt(query.approval_status.as_deref()) {
        builder.push(" AND approval_status = ");
        builder.push_bind(approval_status);
    }

    builder.push(" ORDER BY expense_date DESC LIMIT ");
    builder.push_bind(clamp_limit_in_range(query.limit, 1, 2000));

    let rows = builder.build().fetch_all(pool).await.map_err(|error| {
        tracing::error!(error = %error, "Database query failed");
        AppError::Dependency("External service request failed.".to_string())
    })?;

    Ok(rows
        .into_iter()
        .filter_map(|row| row.try_get::<Option<Value>, _>("row").ok().flatten())
        .collect())
}

async fn attach_reservation_unit_property(
    pool: &sqlx::PgPool,
    org_id: &str,
    reservation_id: &str,
    target: &mut Map<String, Value>,
) -> AppResult<()> {
    let reservation = get_row(pool, "reservations", reservation_id, "id").await?;
    if value_str(&reservation, "organization_id") != org_id {
        return Err(AppError::BadRequest(
            "reservation_id does not belong to this organization.".to_string(),
        ));
    }

    let unit_id = value_str(&reservation, "unit_id");
    if unit_id.is_empty() {
        return Err(AppError::BadRequest(
            "reservation_id is missing unit_id.".to_string(),
        ));
    }
    target.insert("unit_id".to_string(), Value::String(unit_id.clone()));

    let unit = get_row(pool, "units", &unit_id, "id").await?;
    if value_str(&unit, "organization_id") != org_id {
        return Err(AppError::BadRequest(
            "reservation unit does not belong to this organization.".to_string(),
        ));
    }

    if let Some(property_id) = string_from_value(unit.get("property_id")).filter(|v| !v.is_empty())
    {
        target.insert("property_id".to_string(), Value::String(property_id));
    }

    Ok(())
}

fn db_pool(state: &AppState) -> AppResult<&sqlx::PgPool> {
    state.db_pool.as_ref().ok_or_else(|| {
        AppError::Dependency(
            "Supabase database is not configured. Set SUPABASE_DB_URL or DATABASE_URL.".to_string(),
        )
    })
}

fn value_str(row: &Value, key: &str) -> String {
    row.as_object()
        .and_then(|obj| obj.get(key))
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(ToOwned::to_owned)
        .unwrap_or_default()
}

fn string_from_map(map: &Map<String, Value>, key: &str) -> Option<String> {
    string_from_value(map.get(key))
}

fn string_from_value(value: Option<&Value>) -> Option<String> {
    value
        .and_then(Value::as_str)
        .map(str::trim)
        .map(ToOwned::to_owned)
}

fn non_empty_opt(value: Option<&str>) -> Option<String> {
    value
        .map(str::trim)
        .filter(|item| !item.is_empty())
        .map(ToOwned::to_owned)
}
